src/shell/shell_cmds.cpp

```diff
@@ -608,6 +614,61 @@ static std::vector<int> calc_column_widths(const std::vector<int> &word_widths,
 	return col_widths;
 }
 
+char buffer[15] = {0};
+char *format_date(uint16_t year, uint8_t month, uint8_t day)
+{
+	char formatstring[6];
+	char date_format = dos.tables.country[0];
+	char date_separator = dos.tables.country[11];
+	sprintf(formatstring,
+	        date_format == 1 ? "D%cM%cY"
+	                         : (date_format == 2 ? "Y%cM%cD" : "M%cD%cY"),
+	        date_separator, date_separator);
+	Bitu bufferptr = 0;
+	for (Bitu i = 0; i < 5; i++) {
+		if (i == 1 || i == 3) {
+			buffer[bufferptr] = formatstring[i];
+			bufferptr++;
+		} else {
+			if (formatstring[i] == 'M')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%02u", month);
+			if (formatstring[i] == 'D')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%02u", day);
+			if (formatstring[i] == 'Y')
+				bufferptr += (Bitu)sprintf(buffer + bufferptr,
+				                           "%04u", year);
```

Comment:
1. maybe the `index` variable could be an `int` which matches the return-type of `sprintf`; and then these casts would not be needed.

2. `sprintf` returns a negative value if it fails, and casting that to unsigned 64-bit would wrap around and create a large out-of-bounds memory access in `buffer[/2^64-1/]`.

3. Probably need error checking on `sprint`'s return code. 

4. Also consider `safe_sprintf`, which can be used when the buffer is a literal array (and not a pointer)

Reply:
Done 1-3. For 4, I checked that safe_sprintf may be applied directly in this case.